Readme.txt
Squeak Through 1.1.8
====================

Version 1.1.8 was released December 17, 2017, and was tested using Factorio v0.16.5. It primarily features code by Nommy, with contribution from Lupin, and is maintained by Supercheese.

This small mod reduces the collision boxes for many structures, allowing you to "squeak through" them while walking around your base.
Say goodbye to the frustration of having your path blocked by your steam engines or solar panels when walking around your base!

You can toggle the collision box shrinking for each type of entity using the in-game mod settings. Altering these settings will require you to restart the game.


See also the associated forum thread: http://www.factorioforums.com/forum/viewtopic.php?f=91&t=16476


Changelog.txt
---------------------------------------------------------------------------------------------------
Version: 1.1.8
Date: 17. 12. 2017
  Changes:
    - Updated for Factorio 0.16.
---------------------------------------------------------------------------------------------------
Version: 1.1.7
Date: 25. 4. 2017
  Changes:
    - Updated for Factorio 0.15.
    - Added relevant new 0.15 entity-types to config.lua.
---------------------------------------------------------------------------------------------------
Version: 1.1.6
Date: 20. 12. 2016
  Changes:
    - Tree collision boxes are now smaller by default.
---------------------------------------------------------------------------------------------------
Version: 1.1.5
Date: 16. 10. 2016
  Changes:
    - Added exclusions to make the Reactors mod compatible.
    - Switched to data-updates.lua instead of data-final-fixes.lua.
---------------------------------------------------------------------------------------------------
Version: 1.1.4
Date: 27. 8. 2016
  Changes:
    - Updated mod for Factorio 0.14.
---------------------------------------------------------------------------------------------------
Version: 1.1.3
Date: 30. 6. 2016
  Changes:
    - Updated mod for Factorio 0.13.
---------------------------------------------------------------------------------------------------
Version: 1.1.2
Date: 20. 5. 2016
  Changes:
    - Added exclusions to make Logistics Railway compatible.
---------------------------------------------------------------------------------------------------
Version: 1.1.1
Date: 21. 4. 2016
  Changes:
    - Moved config stuff out of data-final-fixes.lua into new config.lua file.
    - Added ability to exclude prototypes by name or type conditionally in config.lua for mod compatibility.
    - Added exclusions to make Nuclear Power and old versions of Rail Tanker compatible.
    - Re-enabled "pump" shrinking (is disabled when Rail Tanker is loaded now).
---------------------------------------------------------------------------------------------------
Version: 1.1.0
Date: 10. 4. 2016
  Major Features:
    - Removed original (extremely simple) code and incorporated the more advanced code by Nommy: https://forums.factorio.com/viewtopic.php?p=114341#p114341
    - Also incorporated Lupin's solution to some incompatibility issues with other mods: https://forums.factorio.com/viewtopic.php?p=146721#p146721
    - This means that all structures of the following prototypes are affected by the mod now, including both vanilla and modded structures:
    - "solar-panel"
    - "accumulator"
    - "generator"
    - "pipe"
    - "pipe-to-ground"
    - "container"
    - "smart-container"
    - "logistic-container"
    - "assembling-machine"
    - "arithmetic-combinator"
    - "decider-combinator"
    - "constant-combinator"
    - "boiler"
    - "electric-pole"
    - "mining-drill"
    - "radar"
    - "storage-tank"
    - "tree"
    - "turret"
    - "beacon"
    - "furnace"
    - "lab"
---------------------------------------------------------------------------------------------------
Version: 1.0.1
Date: 22. 3. 2016
  Changes:
    - Removed the Small Pump's collision box change due to compatibility issues with other mods.
---------------------------------------------------------------------------------------------------
Version: 1.0.0 - First public release
Date: 20. 9. 2015
  Major Features:
    - Initial release, altering the collision boxes of the following structures:
    - *Solar Panels
    - *Accumulators
    - *Steam Engines
    - *Pipes
    - *Chests
    - *Chemical Plants
    - *Oil Refineries
    - *Combinators
    - *Boilers
    - *Substations
    - *Mining Drills
    - *Pumpjacks
    - *Small Pumps
    - *Radar
    - *Fluid Storage Tanks


data.lua
data.raw["solar-panel"]["solar-panel"].collision_box = {{-1.25, -1.25}, {1.25, 1.25}}

data.raw["accumulator"]["basic-accumulator"].collision_box = {{-0.75, -0.75}, {0.75, 0.75}}

data.raw["generator"]["steam-engine"].collision_box = {{-1.25, -2.25}, {1.25, 2.25}}

data.raw["pipe"]["pipe"].collision_box = {{-0.08, -0.08}, {0.08, 0.08}}

data.raw["pipe-to-ground"]["pipe-to-ground"].collision_box = {{-0.08, -0.08}, {0.08, 0.08}}

data.raw["container"]["wooden-chest"].collision_box = {{-0.25, -0.25}, {0.25, 0.25}}

data.raw["container"]["iron-chest"].collision_box = {{-0.25, -0.25}, {0.25, 0.25}}

data.raw["container"]["steel-chest"].collision_box = {{-0.25, -0.25}, {0.25, 0.25}}

data.raw["smart-container"]["smart-chest"].collision_box = {{-0.25, -0.25}, {0.25, 0.25}}

data.raw["logistic-container"]["logistic-chest-active-provider"].collision_box = {{-0.25, -0.25}, {0.25, 0.25}}

data.raw["logistic-container"]["logistic-chest-passive-provider"].collision_box = {{-0.25, -0.25}, {0.25, 0.25}}

data.raw["logistic-container"]["logistic-chest-requester"].collision_box = {{-0.25, -0.25}, {0.25, 0.25}}

data.raw["logistic-container"]["logistic-chest-storage"].collision_box = {{-0.25, -0.25}, {0.25, 0.25}}

data.raw["assembling-machine"]["chemical-plant"].collision_box = {{-1.25, -1.25}, {1.25, 1.25}}

data.raw["assembling-machine"]["oil-refinery"].collision_box = {{-2.25, -2.25}, {2.25, 2.25}}

data.raw["arithmetic-combinator"]["arithmetic-combinator"].collision_box = {{-0.25, -0.75}, {0.25, 0.75}}

data.raw["decider-combinator"]["decider-combinator"].collision_box = {{-0.25, -0.75}, {0.25, 0.75}}

data.raw["constant-combinator"]["constant-combinator"].collision_box = {{-0.25, -0.25}, {0.25, 0.25}}

data.raw["boiler"]["boiler"].collision_box = {{-0.08, -0.08}, {0.08, 0.08}}

data.raw["electric-pole"]["substation"].collision_box = {{-0.75, -0.75}, {0.75, 0.75}}

data.raw["mining-drill"]["burner-mining-drill"].collision_box = {{-0.75, -0.75}, {0.75, 0.75}}

data.raw["mining-drill"]["basic-mining-drill"].collision_box = {{-1.25, -1.25}, {1.25, 1.25}}

data.raw["mining-drill"]["pumpjack"].collision_box = {{-1.25, -1.25}, {1.25, 1.25}}

data.raw["radar"]["radar"].collision_box = {{-1.25, -1.25}, {1.25, 1.25}}

data.raw["storage-tank"]["storage-tank"].collision_box = {{-1.25, -1.25}, {1.25, 1.25}}



data-final-fixes.lua
require "config"

-- Prototype names and types to not alter.
local excluded_prototype_names = {}
local excluded_prototype_types = {}


-- Helper functions to check if a prototype name or type is excluded.
local function prototype_name_excluded(name)
	return excluded_prototype_names[name]
end
local function prototype_type_excluded(type)
	return excluded_prototype_types[type]
end


-- Returns true when an exclusion should be applied or false otherwise.
local function exclusion_applies(exclusion)

	-- Exclusions always apply if no apply_when_object_exists is specified.
	if not exclusion.apply_when_object_exists then
		return true
	else
		-- Exclusions apply when the apply_when_object_exists object actually exists.
		if data.raw[exclusion.apply_when_object_exists.type][exclusion.apply_when_object_exists.name] then
			return true
		end
	end
	
	-- The apply_when_object_exists object didn't exist (the mod which the exclusion was for is not active).
	return false
end


-- Update the exclusion arrays by parsing an exclusion (from config.lua).
local function apply_exclusion(exclusion)
	
	if exclusion.excluded_prototype_names then
		for _,n in pairs(exclusion.excluded_prototype_names) do
			excluded_prototype_names[n] = true
		end
	end
	
	if exclusion.excluded_prototype_types then
		for _,t in pairs(exclusion.excluded_prototype_types) do
			excluded_prototype_types[t] = true
		end
	end
end


-- Parse the exclusions defined in config.lua only applying those which are applicable.
local function apply_exclusions()
	for _,e in pairs(exclusions) do
		if exclusion_applies(e) then
			apply_exclusion(e)
		end
	end
end


-- Returns a coordinate reduced where required to form the specified gap between it and the tile boundary.
local function adjust_coordinate_to_form_gap(coordinate, required_gap)
	
	-- Treat all coordinates as positive to simplify calculations.
	local is_negative_coordinate = (coordinate < 0)
	if is_negative_coordinate then
		coordinate = coordinate * -1
	end
	
	local tile_width = 0.5
	
	-- Calculate the existing gap (how much space there is to the next tile edge or 0 when the coordinate lies on a tile edge).
	local distance_past_last_tile_edge = coordinate % tile_width -- This is how far the collision box extends over any tile edge, and should be 0 for a perfect fit.
	local existing_gap = 0
	if distance_past_last_tile_edge > 0 then
		existing_gap = (tile_width - distance_past_last_tile_edge)
	end
	
	-- Reduce the coordinate to make the gap large enough if it is not already.
	if existing_gap < required_gap then
		coordinate = coordinate + existing_gap - required_gap
		if coordinate < 0 then
			coordinate = 0
		end
	end
	
	-- Make the coordinate negative again if it was originally negative.
	if is_negative_coordinate then
		coordinate = coordinate * -1
	end
	
	return coordinate
end


-- Checks all existing prototypes listed in prototype_type_gap_requirements and reduces their collision box to make a gap large enough to walk though if it is not already.
local function adjust_collision_boxes()
	for prototype_type, required_gap in pairs(prototype_type_gap_requirements) do
	
		-- Don't shrink prototypes of this type if they've been excluded. 
		if not prototype_type_excluded(prototype_type) then
			for prototype_name, prototype in pairs(data.raw[prototype_type]) do
			
				-- If the prototype is not excluded and has a collision box then resize it.
				if (not prototype_name_excluded(prototype_name)) and prototype["collision_box"] then
				
					for y=1,2 do
						for x=1,2 do
							prototype.collision_box[x][y] = adjust_coordinate_to_form_gap(prototype.collision_box[x][y], required_gap)
						end
					end
					
				end
			end
		end
	end
end

-- Exclude prototypes from alteration for mods that have compatibility issues with modified collision boxes.
apply_exclusions()

-- Make the adjustments.
adjust_collision_boxes()


config.lua

-- Note: Any mods which define the collision box size of prototypes in data-final-fixes.lua will not be affected by this code.
--       Prototypes normally defined in data.lua - or data-updates.lua loaded before this - will be properly updated.

-- Table of prototype types to check and the minimum gap required to allow walking in between them. Tiles are 0.5 wide. Can be set to a large value (e.g 1000) create the smallest possible collision box {{0,0},{0,0}}.
prototype_type_gap_requirements =
{
	["solar-panel"]             = 0.25,
	["accumulator"]             = 0.25,
	["generator"]               = 0.25,
	["pipe"]                    = 0.42,
	["pipe-to-ground"]          = 0.42,
	["heat-pipe"]		        = 0.25,
	["reactor"]			        = 0.48,
	["programmable-speaker"]    = 0.25,
	["container"]               = 0.25,
	["logistic-container"]      = 0.25,
	["assembling-machine"]      = 0.25,
	["arithmetic-combinator"]   = 0.25,
	["decider-combinator"]      = 0.25,
	["constant-combinator"]     = 0.25,
	["boiler"]                  = 0.42,
	["electric-pole"]           = 0.25,
	["mining-drill"]            = 0.25,
	["pump"]                    = 0.42,
	["radar"]                   = 0.25,
	["storage-tank"]            = 0.25,
	["tree"]                    = 0.42,
	["turret"]                  = 0.25,
	["beacon"]                  = 0.25,
	["furnace"]                 = 0.25,
	["lab"]                     = 0.25
}
-- All defined prototypes of the types listed above will be checked in adjust_collision_boxes() and have their collision boxes reduced to form the specified gap where needed.



--[[  Defines which prototypes squeak though should not alter for mod compatibility.

		apply_when_object_exists (optional):   Exclusions will only be applied when an object of the specified type and name is found ( in data.raw[type][name] ).
		                                       If omitted the exclusions will always be applied.
			type                                Note: The type and name code can be copied directly from the prototype definition in the mod to here.
			name

		excluded_prototype_names (optional):   (array) Individual prototypes which should not be altered e.g "small-pump" or "steam-engine".

		excluded_prototype_types (optional):   (array) Types of prototypes which should not be altered e.g "pump" or "generator".
		                                       All prototypes from all mods of the types specified will be excluded from alteration.
		                                       Meaning if you exclude "pump" no pumps will have their collision boxes changed.
--]]
exclusions =
{
	{  -- General Exclusions (always applied)
		excluded_prototype_names = {
			-- List names of individual prototypes to always exclude here (if needed).
		}
	},

	{  -- UraniumPower mod    (some test-entities display errors if altered so are excluded below)
		apply_when_object_exists = {
			type = "storage-tank",
			name = "lube-port"
		},
		excluded_prototype_names = {
			"lube-port",
			"fluid-box-1",
			"fluid-box-2"
		}
	},

	-- {  -- Reactors mod
		-- apply_when_object_exists = {
			-- type = "assembling-machine",
			-- name = "nuclear-reactor",
		-- },
		-- excluded_prototype_names = {
			-- "nuclear-reactor",
			-- "reactor-interface",
		-- }
	-- },

}
